src/tla2528.cpp

```diff
@@ -0,0 +1,186 @@
+#include <bitset>
+#include <libhal-util/bit.hpp>
+#include <libhal-util/i2c.hpp>
+#include <libhal/error.hpp>
+#include <libhal/units.hpp>
+#include <libhal-expander/tla2528.hpp>
+
+namespace {
+enum op_codes : hal::byte
+{
+  single_register_read = 0b0001'0000,
+  single_register_write = 0b0000'1000,
+  set_bit = 0b0001'1000,
+  clear_bit = 0b0010'0000,
+  // Continuously reads data from a group of registers. Provide the first
+  // address to read from, if it runs out of valid addresses to read, it
+  // returns zeros. (See Figure 30 on datasheet)
+  continuous_register_read = 0b0011'0000,
+  // Continuously writes data to a group of registers. Provide the first
+  // address to write to.The data sent will automatically write the data to
+  // the next register in ascending order. (See Figure 32 on datasheet)
+  continuous_register_write = 0b0010'1000
+};
+
+enum register_addresses : hal::byte
+{
+  system_status = 0x0,
+  general_cfg = 0x1,
+  data_cfg = 0x2,
+  osr_cfg = 0x3,
+  opmode_cfg = 0x4,
+  pin_cfg = 0x5,
+  gpio_cfg = 0x7,
+  gpo_drive_cfg = 0x9,
+  gpo_value = 0xB,
+  gpi_value = 0xD,
+  sequence_cfg = 0x10,
+  channel_sel = 0x11,
+  auto_seq_ch_sel = 0x12
+};
+}  // namespace
+namespace hal::expander {
+
+tla2528::tla2528(hal::i2c& p_i2c, hal::byte p_i2c_address)
+  : m_i2c_bus(p_i2c)
+  , m_i2c_address(p_i2c_address)
+{
+  // TODO: reset command
+}
+
+void tla2528::set_analog_channel(hal::byte p_channel)
+{
+  throw_if_invalid_channel(p_channel);
+  if (p_channel == m_channel) {
+    return;
+  }
+  std::array<hal::byte, 3> cmd_buffer = { op_codes::single_register_write,
+                                          register_addresses::channel_sel,
+                                          p_channel };
+  hal::write(m_i2c_bus, m_i2c_address, cmd_buffer);
+}
+
+void tla2528::set_pin_mode(pin_mode p_mode, hal::byte p_channel)
+{
+  throw_if_invalid_channel(p_channel);
+  std::array<hal::byte, 5> data_buffer;
+  std::array<hal::byte, 2> read_cmd_buffer = {
+    op_codes::continuous_register_read, register_addresses::pin_cfg
+  };
+  hal::write_then_read(m_i2c_bus, m_i2c_address, read_cmd_buffer, data_buffer);
+
+  hal::byte pin_cfg_reg = data_buffer[0];
+  hal::byte gpio_cfg_reg = data_buffer[2];
+  hal::byte gpo_drive_cfg_reg = data_buffer[4];
+
+  hal::bit_mask channel_mask = hal::bit_mask::from(p_channel);
+  if (p_mode == pin_mode::digital_output_push_pull ||
+      p_mode == pin_mode::digital_output_open_drain) {
+    if (hal::bit_extract(channel_mask, m_object_created) &&
+        !(hal::bit_extract(channel_mask, pin_cfg_reg) ||
+          hal::bit_extract(channel_mask, gpio_cfg_reg))) {
```

Comment:
```suggestion
    if (hal::bit_extract(channel_mask, m_object_created) &&
        !(hal::bit_extract(channel_mask, pin_cfg_reg) ||
          hal::bit_extract(channel_mask, gpio_cfg_reg))) {
```

This logic is hard to follow. The `m_object_created` makes sense, that this is checking if the adaptor has been created but the other flags are not obvious. What you can do is break these up into bools that represent what you mean to express here. Something like

```C++
bool object_already_created = hal::bit_extract(channel_mask, m_object_created);
// etc for what the other ones mean.
bool misconfigured = not hal::bit_extract(channel_mask, pin_cfg_reg) || hal::bit_extract(channel_mask, gpio_cfg_reg);
if (object_already_created && misconfigured) { /* ... */ }
```

Something like that.

Reply:
Tried to do what you suggested, look it over.

Reply:
approved in-person